src/function/algebra/simplify.js

```diff
@@ -454,91 +576,144 @@ export const createSimplify = /* #__PURE__ */ factory(name, dependencies, (
     return new SymbolNode('_p' + _lastsym++)
   }
 
-  function mapRule (nodes, rule) {
+  function mapRule (nodes, rule, context) {
+    let resNodes = nodes
     if (nodes) {
       for (let i = 0; i < nodes.length; ++i) {
-        nodes[i] = applyRule(nodes[i], rule)
+        const newNode = applyRule(nodes[i], rule, context)
+        if (newNode !== nodes[i]) {
+          if (resNodes === nodes) {
+            resNodes = nodes.slice()
+          }
+          resNodes[i] = newNode
```

Comment:
I see `resNodes[i] = newNode` here, and earlier `let resNodes = nodes`. So I suppose it is possible that this line will alter the original `nodes` argument. I think it would be good to do a shallow copy from `nodes` into `resNodes` to prevent altering the input arguments.

Reply:
It is important that `mapRule` not modify `nodes` (i.e., that it returns the === list to nodes) if the rule doesn't alter any of the individual nodes in `nodes`. That's why the code checks each element in turn and only does the shallow copy (via `.slice()` when and if the rule changes one of them. But the shallow copy is already guaranteed to happen if necessary.